docs: polish readme (#231)
OmniSafeAI/omnisafe: README.md

--------------------------------------------------------------------------------

OmniSafe is an infrastructural framework designed to accelerate safe reinforcement learning (RL) research.
It provides a comprehensive and reliable benchmark for safe RL algorithms, and also an out-of-box modular toolkit for researchers.
SafeRL intends to develop algorithms that minimize the risk of unintended harm or unsafe behavior.

OmniSafe stands as the inaugural unified learning framework in the realm of safe reinforcement learning, aiming to foster the Growth of SafeRL Learning Community.
The key features of OmniSafe:

> - **Highly Modular Framework.** OmniSafe presents a highly modular framework, incorporating an extensive collection of tens of algorithms tailored for safe reinforcement learning across diverse domains. This framework is versatile due to its abstraction of various algorithm types and well-designed API interfaces, using the Adapter and Wrapper design components to bridge gaps and enable seamless interactions between different components. This design allows for easy extension and customization, making it a powerful tool for developers working with different types of algorithms.

Reviewer: ```suggestion
- **Highly Modular Framework.** OmniSafe presents a highly modular framework, incorporating an extensive collection of tens of algorithms tailored for safe reinforcement learning across diverse domains. This framework is versatile due to its abstraction of various algorithm types and well-designed API interfaces, using the Adapter and Wrapper design components to bridge gaps and enable seamless interactions between different components. This design allows for easy extension and customization, making it a powerful tool for developers working with different types of algorithms.
```
I am considering `API = Application Programming Interface` and then `API interfaces = Application Programming Interface interfaces`. Is this a repeat?

Reviewer 2: Maybe a better way is:
```suggestion
- **Highly Modular Framework.** OmniSafe presents a highly modular framework, incorporating an extensive collection of tens of algorithms tailored for safe reinforcement learning across diverse domains. This framework is versatile due to its abstraction of various algorithm types and well-designed command line interfaces, using the ``Adapter`` and ``Wrapper`` design components to bridge gaps and enable seamless interactions between different components. This design allows for easy extension and customization, making it a powerful tool for developers working with different types of algorithms.
```

Reviewer 3: This is in Markdown format rather than reStructuredText. You should use

```markdown
`inline code`
```

rather than

```markdown
``inline code``
```

Author: > I am considering `API = Application Programming Interface` and then `API interfaces = Application Programming Interface interfaces`. Is this a repeat?

I will modify it.

--------------------------------------------------------------------------------

> OmniSafe is an infrastructural framework designed to accelerate safe reinforcement learning (RL) research by providing a comprehensive and reliable benchmark for safe RL algorithms.

Reviewer: ```suggestion
OmniSafe is an infrastructural framework designed to accelerate safe reinforcement learning (RL) research by providing a comprehensive and reliable benchmark for safe RL algorithms, out-of-box modular toolkits for researchers.
```

> Safe RL intends to develop algorithms that minimize the risk of unintended harm or unsafe behavior.

Reviewer: ```suggestion
SafeRL intends to develop algorithms that minimize the risk of unintended harm or unsafe behavior.
```
SafeRL or Safe RL maybe can be unified in all our statements.

The key features of OmniSafe:

> - `Highly Modular Framework.` OmniSafe presents a highly modular framework, incorporating an extensive collection of over 40 algorithms tailored for safe reinforcement learning across diverse domains. The framework achieves versatility through the utilization of algorithm-level abstraction and API interfaces, employing two distinctive design components: Adapter and Wrapper.

Reviewer: ```suggestion
- `Highly Modular Framework.` OmniSafe presents a highly modular framework, incorporating an extensive collection of over 30 algorithms tailored for safe reinforcement learning across diverse domains. The framework achieves versatility through the utilization of algorithm-level abstraction and API interfaces, employing two distinctive design components: Adapter and Wrapper.
```
corresponding to the statement in paper

Author: Paper will delete the number, and algorithm table is 40+

> - `Highly Modular Framework.` OmniSafe presents a highly modular framework, incorporating an extensive collection of tens of algorithms tailored for safe reinforcement learning across diverse domains. The framework achieves versatility through the utilization of algorithm-level abstraction and API interfaces, employing two distinctive design components: Adapter and Wrapper.

Reviewer: What is algorithm-level abstraction?

> - `Highly Modular Framework.` OmniSafe presents a highly modular framework, incorporating an extensive collection of tens of algorithms tailored for safe reinforcement learning across diverse domains. The framework achieves versatility through the utilization of algorithm-level abstraction and API interfaces, employing two distinctive design components: Adapter and Wrapper.

Reviewer 3: ```suggestion
- **Highly Modular Framework.** OmniSafe presents a highly modular framework, incorporating an extensive collection of tens of algorithms tailored for safe reinforcement learning across diverse domains. The framework achieves versatility through the utilization of algorithm-level abstraction and API interfaces, employing two distinctive design components: Adapter and Wrapper.
```

> - `Highly Modular Framework.` OmniSafe presents a highly modular framework, incorporating an extensive collection of tens of algorithms tailored for safe reinforcement learning across diverse domains. The framework achieves versatility through the utilization of algorithm-level abstraction and API interfaces, employing two distinctive design components: Adapter and Wrapper.

Reviewer: ```suggestion
- `Highly Modular Framework.` OmniSafe presents a highly modular framework, incorporating an extensive collection of tens of algorithms tailored for safe reinforcement learning across diverse domains. This framework is versatile due to its abstraction of various algorithm types and well-designed API interfaces, using the Adapter and Wrapper design components to bridge gaps and enable seamless interactions between different components. This design allows for easy extension and customization, making it a powerful tool for developers working with different types of algorithms.
```
Will this be better？

Author: I think this will be better.

- `High-performance parallel computing acceleration.` OmniSafe not only supports environment-level asynchronous parallelism but also incorporates agent asynchronous learning. By harnessing the capabilities of Torch.distributed, OmniSafe accelerates the learning process of algorithms by creating unique agents within individual threads.

> - `Fostering the Growth of SafeRL Learning Communities.` OmniSafe stands as the inaugural unified learning framework in the realm of safe reinforcement learning. The platform offers exhaustive API documentation, encompassing user guides elucidating both fundamental and advanced features, illustrative examples, and best practices for algorithm parameter selection and result monitoring.

Reviewer: ```suggestion
- `Fostering the Growth of SafeRL Learning Community.` OmniSafe stands as the inaugural unified learning framework in the realm of safe reinforcement learning. The platform offers exhaustive API documentation, encompassing user guides elucidating both fundamental and advanced features, illustrative examples, and best practices for algorithm parameter selection and result monitoring.
```

> 

Reviewer: ```suggestion
- `Out-of-box toolkits` OmniSafe offers a comprehensive suite of modular toolkits designed to facilitate various tasks such as training, benchmarking, analyzing, and rendering. These toolkits are versatile and can be easily customized to meet the specific needs of users. Furthermore, OmniSafe provides well-written tutorials that enable beginners to quickly get started with the toolkits, while average users can utilize user-friendly APIs to streamline their workflow. Advanced researchers can also take advantage of these toolkits to enhance their research efficiency, without the need to deal with complex and time-consuming code.
```

![Train video](./images/train.gif)

--------------------------------------------------------------------------------

### Table of Contents  <!-- omit in toc --> <!-- markdownlint-disable heading-increment -->
- [🚀Quick Start](#quick-start)
  - [🔥Installation](#installation)
    - [👀Prerequisites](#prerequisites)
    - [✨Install from source](#install-from-source)
    - [✨Install from PyPI](#install-from-pypi)
- [🍪Implemented Algorithms](#implemented-algorithms)
  - [🍩Examples](#examples)
    - [🏠Algorithms Registry](#algorithms-registry)
    - [☁️Supported Environments](#️supported-environments)
    - [🚢Try with CLI](#try-with-cli)
- [🍷Getting Started](#getting-started)
  - [🥢Important Hints](#important-hints)
  - [🍴Quickstart: Colab on the Cloud](#quickstart-colab-on-the-cloud)
- [🔖Changelog](#changelog)
- [⚓The OmniSafe Team](#the-omnisafe-team)

> - [📖License](#license)

Reviewer 3: I still think we should not add emojis to section headers.

Author: Why

- [X] **[ICML 2021]** [CRPO: A New Approach for Safe Reinforcement Learning with Convergence Guarantee](https://arxiv.org/abs/2011.05869)
- [x] **[IJCAI 2022]** [Penalized Proximal Policy Optimization for Safe Reinforcement Learning(P3O)](https://arxiv.org/pdf/2205.11814.pdf)

> <summary><b><big>Off Policy SafeRL</big></b></summary>

Reviewer: ```suggestion
<summary><b><big>Off Policy Safe</big></b></summary>
```

Reviewer: or change `Model-Based Safe` to `Model-Based SafeRL`?

Author: I think change Model-Based Safe to Model-Based SafeRL will be better.

<summary><b><big>Others</big></b></summary>

- [X] [Safe Exploration in Continuous Action Spaces (Safety Layer)](https://arxiv.org/abs/1801.08757)
- [ ] **[RA-L 2021]** [Recovery RL: Safe Reinforcement Learning with Learned Recovery Zones](https://arxiv.org/abs/2010.15920)
- [X] **[ICML 2022]** [Sauté RL: Almost Surely Safe Reinforcement Learning Using State Augmentation (SauteRL)](https://arxiv.org/abs/2202.06558)
- [X] **[NeurIPS 2022]** [Effects of Safety State Augmentation on Safe Exploration](https://arxiv.org/abs/2206.02675)

</details>

--------------------------------------------------------------------------------

### 🍩Examples

```bash
cd examples
python train_policy.py --algo PPOLag --env-id SafetyPointGoal1-v0 --parallel 1 --total-steps 1024000 --device cpu --vector-env-nums 1 --torch-threads 1
```

#### 🏠Algorithms Registry
<table>
<thead>
  <tr>
    <th>Domains</th>
    <th>Types</th>
    <th>Algorithms Registry</th>
  </tr>
</thead>
<tbody>
  <tr>
    <td rowspan="4">On Policy</td>
    <td rowspan="2">Primal Dual</td>
    <td>TRPOLag; PPOLag; PDO; RCPO; OnCRPO</td>
  </tr>
  <tr>

>     <td>TRPOPIDLag; PPOPIDLag;CPPOPID</td>

Reviewer: ```suggestion
    <td>TRPOPIDLag; PPOPIDLag; CPPOPID</td>
```

  </tr>
  <tr>
    <td rowspan="3">Model-based</td>
    <td>Online Plan</td>
    <td>SafeLOOP; CCEPETS; RCEPETS</td>
  </tr>
  <tr>
    <td><span style="font-weight:400;font-style:normal">Pessimistic Estimate</span></td>
    <td>LA-MBDA; CAPPETS</td>
  </tr>
  <tr>
    <td>Imaginary Train</td>
    <td>SMBPO; MBPPOLag</td>
  </tr>
  <tr>
    <td rowspan="2">Control</td>
    <td>Recovery/ Optimal Layer</td>
    <td>SafetyLayer; RecoveryRL</td>
  </tr>
  <tr>
    <td>Lyapunov</td>
    <td>SPPO; SPPOM</td>
  </tr>
  <tr>
    <td rowspan="2">Offline</td>
    <td>Q-Learning Based</td>
    <td>BCQLag; C-CRR</td>
  </tr>
  <tr>
    <td>DICE Based</td>
    <td>COptDICE</td>
  </tr>
  <tr>
    <td rowspan="3">Other Formulation MDP</td>
    <td>ET-MDP</td>
    <td><span style="font-weight:400;font-style:normal">PPO</span>EarlyTerminated; TRPOEarlyTerminated</td>
  </tr>
  <tr>
    <td>SauteRL</td>
    <td>PPOSaute; TRPOSaute</td>
  </tr>
  <tr>
    <td>SimmerRL</td>
    <td><span style="font-weight:400;font-style:normal">PPOSimmerPID; TRPOSimmerPID</span></td>
  </tr>
</tbody>
</table>

#### ☁️Supported Environments

> Environment id in [Safety Gymnasium](https://www.safety-gymnasium.com/), here a list of envs that safety-gymnasium supports.

Reviewer: ```suggestion
Environment id in [Safety-Gymnasium](https://www.safety-gymnasium.com/), here a list of envs that safety-gymnasium supports.
```

python train_policy.py --algo PPOLag --env-id SafetyPointGoal1-v0 --parallel 1 --total-steps 1024000 --device cpu --vector-env-nums 1 --torch-threads 1
```

#### Algorithms Registry

<table>
<thead>
  <tr>
    <th>Domains</th>
    <th>Types</th>
    <th>Algorithms Registry</th>
  </tr>
</thead>
<tbody>
  <tr>
    <td rowspan="4">On Policy</td>
    <td rowspan="2">Primal Dual</td>
    <td>TRPOLag; PPOLag; PDO; RCPO; OnCRPO</td>
  </tr>
  <tr>

>     <td>TRPOPIDLag; PPOPIDLag; CPPOPID</td>

Reviewer 2: no PPOPIDLag, just CPPOPID

  </tr>
  <tr>
    <td>Convex Optimization</td>
    <td><span style="font-weight:400;font-style:normal">CPO; PCPO; </span>FOCOPS; CUP</td>
  </tr>
  <tr>
    <td>Penalty Function</td>
    <td>IPO; P3O</td>
  </tr>
  <tr>
    <td rowspan="2">Off Policy</td>
    <td rowspan="2">Primal-Dual</td>

>     <td>SACLAG; DDPGLag; TD3LAG</td>

Reviewer 2: TD3Lag

> More information about environments, please refer to [Safety Gymnasium](https://www.safety-gymnasium.com/)

Reviewer: ```suggestion
More information about environments, please refer to [Safety-Gymnasium](https://www.safety-gymnasium.com/).
```
